File: Aran/Panda/Arrival/trunk/AranPlugin.vb
Imports Aran.AranEnvironment
Imports Microsoft.Win32
Imports Aran.Aim
Imports System.Collections.Generic

'#Const IncludeVisualManoeuvring = True

Public Class ArrivalAranPlugin
    Inherits AranPlugin

    Const _name As String = "PANDA Conventional - Arrival"

    'Dim _isInit As Boolean
    Dim _tsmiNonPrecArrival As ToolStripMenuItem
    Dim _tsmiPrecArrival As ToolStripMenuItem
    Dim _tsmiInitialArea As ToolStripMenuItem
    Dim _tsmiDeadRekconing As ToolStripMenuItem
    Dim _tsmiVisualManoeuvring As ToolStripMenuItem

#Region "IAranPlugin Members"

	Public Overrides ReadOnly Property Id As Guid
		Get
			Return New Guid("d548ff47-74c9-49ab-aa9a-80161ed94a27")
		End Get
	End Property

	Public Overrides ReadOnly Property Name As String
        Get
            Return _name
        End Get
    End Property

    Public Overrides Function GetLayerFeatureTypes() As List(Of FeatureType)
        Dim featTypes As List(Of FeatureType) = New List(Of FeatureType)()

        featTypes.Add(FeatureType.AirportHeliport)
        featTypes.Add(FeatureType.VerticalStructure)
        featTypes.Add(FeatureType.VOR)
        featTypes.Add(FeatureType.DME)
        featTypes.Add(FeatureType.NDB)
		featTypes.Add(FeatureType.InstrumentApproachProcedure)
		featTypes.Add(FeatureType.DesignatedPoint)
        featTypes.Add(FeatureType.RunwayCentrelinePoint)

        Return featTypes
    End Function

    Public Overrides Sub Startup(aranEnv As AranEnvironment.IAranEnvironment)
        gAranEnv = aranEnv
        gHookHelper = New ESRI.ArcGIS.Controls.HookHelper
        gHookHelper.Hook = gAranEnv.HookObject

        LangCode = 1033

        SetThreadLocale(LangCode)
        Arrival.My.Resources.Culture = New System.Globalization.CultureInfo(LangCode)


        Dim menuItem As New ToolStripMenuItem()
        menuItem.Text = My.Resources.str00080
        _tsmiNonPrecArrival = New ToolStripMenuItem()
        _tsmiNonPrecArrival.Text = My.Resources.str00157  'NonPrecApproach
        _tsmiNonPrecArrival.Tag = 0
        AddHandler _tsmiNonPrecArrival.Click, AddressOf NonPrecArrival_Click
        menuItem.DropDownItems.Add(_tsmiNonPrecArrival)
        _tsmiPrecArrival = New ToolStripMenuItem()
        _tsmiPrecArrival.Text = My.Resources.str00158 'PrecApproach
        _tsmiPrecArrival.Tag = 1
        AddHandler _tsmiPrecArrival.Click, AddressOf PrecArrival_Click
        menuItem.DropDownItems.Add(_tsmiPrecArrival)

        _tsmiInitialArea = New ToolStripMenuItem()
        _tsmiInitialArea.Text = My.Resources.str00159 'InitialArea
        _tsmiInitialArea.Tag = 2
        AddHandler _tsmiInitialArea.Click, AddressOf InitialArea_Click
        menuItem.DropDownItems.Add(_tsmiInitialArea)

        _tsmiDeadRekconing = New ToolStripMenuItem()
        _tsmiDeadRekconing.Text = My.Resources.str00160   'DeadRekconing
        _tsmiDeadRekconing.Tag = 3
        AddHandler _tsmiDeadRekconing.Click, AddressOf DeadRekconing_Click
        menuItem.DropDownItems.Add(_tsmiDeadRekconing)

#If IncludeVisualManoeuvring Then
        'Aran.PANDA.VisualManoeuvring.GlobalVars.gAranEnv = aranEnv
        Aran.PANDA.VisualManoeuvring.GlobalVars.gAranEnv = aranEnv

        _tsmiVisualManoeuvring = New ToolStripMenuItem()
        _tsmiVisualManoeuvring.Text = "Visual Manoeuvring"
        _tsmiVisualManoeuvring.Tag = 4
        AddHandler _tsmiVisualManoeuvring.Click, AddressOf VisualManoeuvring_Click
        menuItem.DropDownItems.Add(_tsmiVisualManoeuvring)
#End If

		aranEnv.AranUI.AddMenuItem(AranMapMenu.Applications, menuItem)
    End Sub
#End Region

    Public Function InitCommands() As Boolean
        Return InitCommand() > 0
    End Function

	Public Sub NonPrecArrival_Click(sender As Object, e As EventArgs)
		If CurrCmd = 1 Then Return
		CurrCmd = 0

		Try
            If (SlotSelector()=False)
                Return
            End If
            NativeMethods.ShowPandaBox(gAranEnv.Win32Window.Handle.ToInt32())

            If InitCommands() Then

				Dim ArrivalFrm As CArrivalFrm = New CArrivalFrm

				If ArrivalFrm.IsDisposed Then
					NativeMethods.HidePandaBox()
					Return
				End If

				CurrCmd = 1
				AddHandler ArrivalFrm.FormClosed, AddressOf AllForms_FormClosed

                'If toolbar Is Nothing OrElse toolbar.IsDisposed Then
                '	toolbar = New ArrivalToolbar
                'End If

                _tsmiNonPrecArrival.Checked = True
                _tsmiPrecArrival.Enabled = False
                _tsmiInitialArea.Enabled = False
                _tsmiDeadRekconing.Enabled = False
#If IncludeVisualManoeuvring Then
                _tsmiVisualManoeuvring.Enabled = False
#End If
				ArrivalFrm.Show(_Win32Window)
				ArrivalFrm.ListView0001.Focus()

			End If

			NativeMethods.HidePandaBox()

		Catch aranEx As AranException
			NativeMethods.HidePandaBox()
			If Not aranEx.Handled Then
				aranEx.ShowMessageBox()
			End If
		Catch exception As Exception
			NativeMethods.HidePandaBox()
            MessageBox.Show("Form initialization error."+exception.Message,"Arrival",MessageBoxButtons.OK,MessageBoxIcon.Error)
		End Try

		'If InitCommands() Then
		'	Dim tsb As ToolStripItem = CType(sender, ToolStripItem)
		'	tsb.Enabled = False

		'	Dim arrivalFrm As CArrivalFrm = New CArrivalFrm
		'	arrivalFrm.Show(s_Win32Window)
		'	arrivalFrm.ListView0001.Focus()

		'	AddHandler arrivalFrm.FormClosed, AddressOf AllForms_FormClosed
		'	arrivalFrm.Tag = tsb
		'End If
	End Sub

	Public Sub PrecArrival_Click(sender As Object, e As EventArgs)
		If CurrCmd = 2 Then Return
		CurrCmd = 0

		Try
            If (SlotSelector()=False)
                Return
            End If
			NativeMethods.ShowPandaBox(gAranEnv.Win32Window.Handle.ToInt32())

			If InitCommands() Then

				Dim PrecisionFrm As CPrecisionFrm = New CPrecisionFrm

				If PrecisionFrm.IsDisposed Then
					NativeMethods.HidePandaBox()
					Return
				End If

				CurrCmd = 2

				AddHandler PrecisionFrm.FormClosed, AddressOf AllForms_FormClosed

				If CommandBar Is Nothing OrElse CommandBar.IsDisposed Then
					CommandBar = New ArrivalToolbar
				End If

				_tsmiNonPrecArrival.Enabled = False
				_tsmiPrecArrival.Checked = True
				_tsmiInitialArea.Enabled = False
				_tsmiDeadRekconing.Enabled = False
#If IncludeVisualManoeuvring Then
                _tsmiVisualManoeuvring.Enabled = False
#End If
				'CommandBar.Show(s_Win32Window)

				Toolbar = CommandBar.Toolbar

				PrecisionFrm.Show(_Win32Window)
				PrecisionFrm.ComboBox0001.Focus()
			End If
			NativeMethods.HidePandaBox()
		Catch ex As Exception
			NativeMethods.HidePandaBox()
			Throw ex 'New Exception("Form initialization error.")
		End Try
	End Sub

	Public Sub InitialArea_Click(sender As Object, e As EventArgs)
		If CurrCmd = 3 Then Return
		CurrCmd = 0

		Try
            If (SlotSelector()=False)
                Return
            End If
			NativeMethods.ShowPandaBox(gAranEnv.Win32Window.Handle.ToInt32())

			If InitCommands() Then
				Dim InitialApproach As InitialApproach = New InitialApproach

				If InitialApproach.IsDisposed Then
					NativeMethods.HidePandaBox()
					Return
				End If

				CurrCmd = 3
				AddHandler InitialApproach.FormClosed, AddressOf AllForms_FormClosed

				_tsmiNonPrecArrival.Enabled = False
				_tsmiPrecArrival.Enabled = False
				_tsmiInitialArea.Checked = True
				_tsmiDeadRekconing.Enabled = False
#If IncludeVisualManoeuvring Then
                _tsmiVisualManoeuvring.Enabled = False
#End If
				InitialApproach.Show(_Win32Window)
				InitialApproach.ComboBox002.Focus()
			End If

			NativeMethods.HidePandaBox()

		Catch ex As Exception

			NativeMethods.HidePandaBox()

            MessageBox.Show("Form initialization error."+ex.Message)
			'Throw ex  'New Exception("Form initialization error.")
		End Try
	End Sub

	Public Sub DeadRekconing_Click(sender As Object, e As EventArgs)
		If CurrCmd = 4 Then Return
		CurrCmd = 0

		Try
            If (SlotSelector()=False)
                Return
            End If
			NativeMethods.ShowPandaBox(gAranEnv.Win32Window.Handle.ToInt32())

			If InitCommands() Then
				Dim InitialDeadApproach As CInitialDeadApproach = New CInitialDeadApproach

				If InitialDeadApproach.IsDisposed Then
					NativeMethods.HidePandaBox()
					Return
				End If

				CurrCmd = 4
				AddHandler InitialDeadApproach.FormClosed, AddressOf AllForms_FormClosed

				_tsmiNonPrecArrival.Enabled = False
				_tsmiPrecArrival.Enabled = False
				_tsmiInitialArea.Enabled = False
				_tsmiDeadRekconing.Checked = True
#If IncludeVisualManoeuvring Then
                _tsmiVisualManoeuvring.Enabled = False
#End If
				InitialDeadApproach.Show(_Win32Window)
				InitialDeadApproach.TextBox014.Focus()
			End If

			NativeMethods.HidePandaBox()

		Catch ex As Exception

			NativeMethods.HidePandaBox()

			Throw ex  'New Exception("Form initialization error.")
		End Try
	End Sub

#If IncludeVisualManoeuvring Then
    'Dim VisualManoeuvring As Aran.PANDA.VisualManoeuvring.Forms.MainForm

    'Public Sub VisualManoeuvring_Click(sender As Object, e As EventArgs)
    '    If CurrCmd = 5 Then Return
    '    CurrCmd = 0

    '    Try
    '        Aran.PANDA.VisualManoeuvring.GlobalVars.InitCommand()
    '        If VisualManoeuvring Is Nothing OrElse VisualManoeuvring.IsDisposed Then
    '            VisualManoeuvring = New Aran.PANDA.VisualManoeuvring.Forms.MainForm
    '        End If

    '        CurrCmd = 5
    '        AddHandler VisualManoeuvring.FormClosed, AddressOf AllForms_FormClosed

    '        _tsmiNonPrecArrival.Enabled = False
    '        _tsmiPrecArrival.Enabled = False
    '        _tsmiInitialArea.Enabled = False
    '        _tsmiDeadRekconing.Enabled = False
    '        _tsmiVisualManoeuvring.Checked = True

    '        VisualManoeuvring.Show(s_Win32Window)
    '        'VisualManoeuvring.NestedControl.Focus()
    '    Catch ex As Exception
    '        Throw New Exception("Form initialization error.")
    '    End Try
    'End Sub


    Dim VisualManoeuvring As Aran.PANDA.VisualManoeuvring.Forms.MainForm

    Public Sub VisualManoeuvring_Click(sender As Object, e As EventArgs)
        If CurrCmd = 5 Then Return
        CurrCmd = 0

        Try
            NativeMethods.ShowPandaBox(gAranEnv.Win32Window.Handle.ToInt32())

            Aran.PANDA.VisualManoeuvring.GlobalVars.InitCommand()
            If VisualManoeuvring Is Nothing OrElse VisualManoeuvring.IsDisposed Then
                VisualManoeuvring = New Aran.PANDA.VisualManoeuvring.Forms.MainForm
            End If

            CurrCmd = 5
            AddHandler VisualManoeuvring.FormClosed, AddressOf AllForms_FormClosed

            _tsmiNonPrecArrival.Enabled = False
            _tsmiPrecArrival.Enabled = False
            _tsmiInitialArea.Enabled = False
            _tsmiDeadRekconing.Enabled = False
            _tsmiVisualManoeuvring.Checked = True

            VisualManoeuvring.Show(s_Win32Window)
            'VisualManoeuvring.NestedControl.Focus()

            NativeMethods.HidePandaBox()

        Catch ex As Exception

            NativeMethods.HidePandaBox()

            Throw ex  'New Exception("Form initialization error.")
        End Try
    End Sub
#End If

	Private Sub AllForms_FormClosed(sender As Object, e As FormClosedEventArgs)
		_tsmiNonPrecArrival.Checked = False
		_tsmiPrecArrival.Checked = False
		_tsmiInitialArea.Checked = False
		_tsmiDeadRekconing.Checked = False
#If IncludeVisualManoeuvring Then
        _tsmiVisualManoeuvring.Checked = False
#End If
		_tsmiNonPrecArrival.Enabled = True
		_tsmiPrecArrival.Enabled = True
		_tsmiInitialArea.Enabled = True
		_tsmiDeadRekconing.Enabled = True

        Toolbar = Nothing

		If Not (CommandBar Is Nothing) Then
			CommandBar.Close()
			CommandBar = Nothing
        End If
        
        Dim formd As System.Windows.Forms.Form= CType(sender, System.Windows.Forms.Form)  
        RemoveHandler formd.FormClosed,AddressOf AllForms_FormClosed  

        sender = Nothing
        GC.Collect()
		
        'If Not CommandBar Is Nothing And Not CommandBar.IsDisposed Then
		'	CommandBar.Dispose()
		'	CommandBar = Nothing
		'End If

		'If Not PrecisionFrm Is Nothing And Not PrecisionFrm.IsDisposed Then
		'	PrecisionFrm.Dispose()
		'End If

#If IncludeVisualManoeuvring Then
        _tsmiVisualManoeuvring.Enabled = True
#End If
		CurrCmd = 0
        

	End Sub
End Class

File: Aran/Panda/RNAV/AdvancedPBN/trunk/AranPlugin.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Aran.AranEnvironment;
using System.Windows.Forms;
using Aran.Aim;
using Aran.Aim.Features;
using Aran.Aim.Enums;
using Aran.Aim.Data;
//using Aran.PANDA.RNAV.SGBAS.Properties;

namespace Aran.PANDA.RNAV.SGBAS
{
	public class DeparturePlugin : AranPlugin
	{
		const string PlaginName = "PANDA RNAV - AdvancedPBN";
		const string GuidString = "5F6AECCF-12EE-4440-A52F-64B099A41046";

		private int _currCmd;
		private Guid _id;
		private ToolStripMenuItem _tsmiSBAS;
		private ToolStripMenuItem _tsmiGBAS;
		private ToolStripMenuItem _tsmiRNP_AR;
		private ToolStripMenuItem _tsmiOASCalc;

		public override Guid Id
		{
			get { return _id; }
		}

		public override string Name
		{
			get { return PlaginName; }
		}

		public override List<FeatureType> GetLayerFeatureTypes()
		{
			var list = new List<FeatureType>();

			list.Add(FeatureType.AirportHeliport);
			list.Add(FeatureType.RunwayCentrelinePoint);
			list.Add(FeatureType.DesignatedPoint);
			list.Add(FeatureType.VOR);
			list.Add(FeatureType.DME);
			list.Add(FeatureType.VerticalStructure);

			return list;
		}

		public static void ShowPandaBox()
		{
			Aran.PANDA.Common.NativeMethods.ShowPandaBox(GlobalVars.gAranEnv.Win32Window.Handle.ToInt32());
		}

		private static void HidePandaBox()
		{
			Aran.PANDA.Common.NativeMethods.HidePandaBox();
		}

		public override void Startup(IAranEnvironment aranEnv)
		{
			GlobalVars.gAranEnv = aranEnv;
			GlobalVars.gAranGraphics = aranEnv.Graphics;

			//GlobalVars.gHookHelper = new ESRI.ArcGIS.Controls.HookHelper();
			//GlobalVars.gHookHelper.Hook = GlobalVars.gAranEnv.HookObject;

			_currCmd = -1;
			_id = new Guid(GuidString);

			_tsmiSBAS = new ToolStripMenuItem();
			_tsmiSBAS.Text = "SBAS";
			_tsmiSBAS.Tag = 0;
			_tsmiSBAS.Click += new EventHandler(cmdSBAS_Click);

			_tsmiGBAS = new ToolStripMenuItem();
			_tsmiGBAS.Text = "GBAS";
			_tsmiGBAS.Tag = 1;
			_tsmiGBAS.Click += new EventHandler(cmdGBAS_Click);

			_tsmiRNP_AR = new ToolStripMenuItem();
			_tsmiRNP_AR.Text = "RNP AR APCH";
			_tsmiRNP_AR.Tag = 2;
			_tsmiRNP_AR.Click += new EventHandler(cmdRNP_AR_Click);

			_tsmiOASCalc = new ToolStripMenuItem();
			_tsmiOASCalc.Text = "OAS Calculator";
			_tsmiOASCalc.Tag = 3;
			_tsmiOASCalc.Click += new EventHandler(cmdOASCalc_Click);

			ToolStripMenuItem menuItem = new ToolStripMenuItem();
			menuItem.Text = "Advanced PBN";
			menuItem.DropDownItems.Add(_tsmiSBAS);
			menuItem.DropDownItems.Add(_tsmiGBAS);
			menuItem.DropDownItems.Add(_tsmiRNP_AR);
			menuItem.DropDownItems.Add(_tsmiOASCalc);

			aranEnv.AranUI.AddMenuItem(AranMapMenu.Applications, menuItem);
		}

		//AltimeterSource
		private void AllForms_FormClosed(object sender, FormClosedEventArgs e)
		{
			_currCmd = -1;
			Toolbar = null;

			_tsmiSBAS.Checked = false;
			_tsmiGBAS.Checked = false;
			_tsmiRNP_AR.Checked = false;
			_tsmiOASCalc.Checked = false;

			_tsmiSBAS.Enabled = true;
			_tsmiGBAS.Enabled = true;
			_tsmiRNP_AR.Enabled = true;
			_tsmiOASCalc.Enabled = true;

			if (GlobalVars.VisibilityBar != null)
			{
				GlobalVars.VisibilityBar.Close();
				GlobalVars.VisibilityBar = null;
			}

			sender = null;
			GC.Collect();
		}

		void cmdSBAS_Click(object sender, EventArgs e)
		{
			if (_currCmd == (int)((ToolStripMenuItem)sender).Tag)
				return;

			_currCmd = -1;

			try
			{
                if (!SlotSelector())
                    return;
				ShowPandaBox();

				_currCmd = (int)((ToolStripMenuItem)sender).Tag;
				GlobalVars.InitCommand();

				if (GlobalVars.VisibilityBar == null || GlobalVars.VisibilityBar.IsDisposed)
					GlobalVars.VisibilityBar = new ToolbarForm();

				Toolbar = GlobalVars.VisibilityBar.RNAVToolStrip;

				SBASForm SBASForm = new SBASForm();

				SBASForm.FormClosed += AllForms_FormClosed;
				SBASForm.Show(GlobalVars.Win32Window);
				SBASForm.ComboBox0001.Focus();


				_tsmiSBAS.Checked = true;
				_tsmiGBAS.Checked = false;
				_tsmiRNP_AR.Checked = false;
				//_tsmiOASCalc.Checked = false;

				_tsmiSBAS.Enabled = false;
				_tsmiGBAS.Enabled = false;
				_tsmiRNP_AR.Enabled = false;
				//_tsmiOASCalc.Enabled = false;

				HidePandaBox();
				SBASForm = null;
			}
			catch(Exception ex)
			{
				HidePandaBox();
				_currCmd = -1;
				MessageBox.Show(GlobalVars.Win32Window, ex.Message, ((ToolStripMenuItem)sender).Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
			}
		}

		void cmdGBAS_Click(object sender, EventArgs e)
		{
			if (_currCmd == (int)((ToolStripMenuItem)sender).Tag)
				return;

			_currCmd = -1;

			try
			{
                if (!SlotSelector())
                    return;
                ShowPandaBox();

				_currCmd = (int)((ToolStripMenuItem)sender).Tag;
				GlobalVars.InitCommand();

				if (GlobalVars.VisibilityBar == null || GlobalVars.VisibilityBar.IsDisposed)
					GlobalVars.VisibilityBar = new ToolbarForm();

				Toolbar = GlobalVars.VisibilityBar.RNAVToolStrip;

				GBASForm GBASForm = new GBASForm();
				GBASForm.FormClosed += AllForms_FormClosed;
				GBASForm.Show(GlobalVars.Win32Window);
				GBASForm.ComboBox0001.Focus();

				_tsmiSBAS.Checked = false;
				_tsmiGBAS.Checked = true;
				_tsmiRNP_AR.Checked = false;
				//_tsmiOASCalc.Checked = false;

				_tsmiSBAS.Enabled = false;
				_tsmiGBAS.Enabled = false;
				_tsmiRNP_AR.Enabled = false;
				//_tsmiOASCalc.Enabled = false;

				HidePandaBox();
				GBASForm = null;
			}
			catch (Exception ex)
			{
				HidePandaBox();
				_currCmd = -1;
				MessageBox.Show(GlobalVars.Win32Window, ex.Message, ((ToolStripMenuItem)sender).Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
			}
		}

		void cmdRNP_AR_Click(object sender, EventArgs e)
		{
			if (_currCmd == (int)((ToolStripMenuItem)sender).Tag)
				return;

			_currCmd = -1;

			try
			{
                if (!SlotSelector())
                    return;
                ShowPandaBox();

				_currCmd = (int)((ToolStripMenuItem)sender).Tag;
				GlobalVars.InitCommand();

				bool havePrimary = false;

				for (int i = 0; i < GlobalVars.CurrADHP.AltimeterSource.Count; i++)
				{
					if (GlobalVars.CurrADHP.pAirportHeliport.AltimeterSource[i] != null)
					{
						Aim.DataTypes.FeatureRef fr = GlobalVars.CurrADHP.pAirportHeliport.AltimeterSource[i].Feature;
						//Aran.Aim.Features.AltimeterSource als = (Aran.Aim.Features.AltimeterSource)fr.GetFeature();
						Aran.Aim.Features.AltimeterSource als = (Aran.Aim.Features.AltimeterSource)Aran.Queries.ExtensionFeature.GetFeature(fr);

						if (als.IsPrimary != null && als.IsPrimary == true)
						{
							havePrimary = true;
							break;
						}
					}
				}

				havePrimary = true;

				if (!havePrimary)
				{
					HidePandaBox();
					_currCmd = -1;
					MessageBox.Show("RNP AR APCH procedure shall not be promulgated for use with remote altimeter setting sources.",
						((ToolStripMenuItem)sender).Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
					return;
				}

				if (GlobalVars.VisibilityBar == null || GlobalVars.VisibilityBar.IsDisposed)
					GlobalVars.VisibilityBar = new ToolbarForm();

				Toolbar = GlobalVars.VisibilityBar.RNAVToolStrip;

				RNP_ARForm RNP_ARForm = new RNP_ARForm();
				RNP_ARForm.FormClosed += AllForms_FormClosed;
				RNP_ARForm.Show(GlobalVars.Win32Window);
				RNP_ARForm.ComboBox0001.Focus();
				//GlobalVars.VisibilityBar.Show(GlobalVars.Win32Window);


				_tsmiSBAS.Checked = false;
				_tsmiGBAS.Checked = false;
				_tsmiRNP_AR.Checked = true;
				//_tsmiOASCalc.Checked = false;

				_tsmiSBAS.Enabled = false;
				_tsmiGBAS.Enabled = false;
				_tsmiRNP_AR.Enabled = false;
				//_tsmiOASCalc.Enabled = false;

				HidePandaBox();
				RNP_ARForm = null;
			}
			catch (Exception ex)
			{
				HidePandaBox();
				_currCmd = -1;
				MessageBox.Show(GlobalVars.Win32Window, ex.Message, ((ToolStripMenuItem)sender).Text, MessageBoxButtons.OK, MessageBoxIcon.Error);
			}
		}

		void cmdOASCalc_Click(object sender, EventArgs e)
		{
			if (_currCmd == (int)((ToolStripMenuItem)sender).Tag)
				return;

			_currCmd = -1;

			try
			{
                if (!SlotSelector())
                    return;
                _currCmd = (int)((ToolStripMenuItem)sender).Tag;

				OASDlg oasDlg = new OASDlg();

				oasDlg.FormClosed += AllForms_FormClosed;
				oasDlg.ShowDialog(GlobalVars.Win32Window);
				oasDlg = null;
			}
			finally
			{
				_currCmd = -1;
			}
		}

        public bool SlotSelector()
        {

            var slotSelector = true;

            var dbProvider = GlobalVars.gAranEnv.DbProvider as DbProvider;
            if (dbProvider != null && dbProvider.ProviderType == DbProviderType.TDB)
            {
                dynamic methodResult = new System.Dynamic.ExpandoObject();
                dbProvider.CallSpecialMethod("SelectSlot", methodResult);
                slotSelector = methodResult.Result;
            }

            if (!slotSelector)
            {
                MessageBox.Show("Please first select slot!", "Slot Selector", MessageBoxButtons.OK, MessageBoxIcon.Warning);
                return slotSelector;
            }
            return slotSelector;
        }
    }
}
